six/modules/c++/six/source/NITFReadControl.cpp

```diff
@@ -127,42 +120,60 @@ DataType NITFReadControl::getDataType(nitf::DESegment& segment)
     }
     else if (desid == Constants::DES_USER_DEFINED_SUBHEADER_TAG)
     {
-        // Check whether DES is SICD/SIDD, or something of a different
-        // sort. Note that we need to check the subheader
-        // length first rather than calling tre.getCurrentSize() in
-        // case there is no subheader (in which case
-        // subheader.getSubheaderFields() will throw with a
-        // NITRO-specific message).
-        const sys::Uint64_T
-            subheaderLength(subheader.getSubheaderFieldsLength());
+        // Check whether DES is SICD/SIDD, or something of a different sort.
         if (subheaderLength !=
             Constants::DES_USER_DEFINED_SUBHEADER_LENGTH)
         {
             return DataType::NOT_SET;
         }
-        nitf::TRE tre = subheader.getSubheaderFields();
-        if (tre.getTag() !=
-            Constants::DES_USER_DEFINED_SUBHEADER_TAG)
+        if (treTag != Constants::DES_USER_DEFINED_SUBHEADER_TAG)
         {
             return DataType::NOT_SET;
         }
-
-        std::string field = tre.getField("DESSHSI").toString();
-        str::trim(field);
-
-        if (field == Constants::SICD_DESSHSI)
+        if (desshsiField == Constants::SICD_DESSHSI)
         {
             return DataType::COMPLEX;
         }
-        else if (field == Constants::SIDD_DESSHSI)
+        else if (desshsiField == Constants::SIDD_DESSHSI)
         {
             return DataType::DERIVED;
         }
-        else
+    }
+    return DataType::NOT_SET;
+}
+
+DataType NITFReadControl::getDataType(nitf::DESegment& segment)
+{
+    // NOTE: Versions of SICD <= 1.1 and SIDD <= 1.0 prefixed FTITLE with
+    //       SICD or SIDD, so for old files we could key off of that. Since
+    //       that's not guaranteed for newer versions though, we now use the
+    //       DESID for really old versions and the DESSHSI from
+    //       XML_DATA_CONTENT for newer versions.
+    nitf::DESubheader subheader = segment.getSubheader();
+    std::string desid = subheader.getTypeID().toString();
+    str::trim(desid);
+
+    // Note that we need to check the subheader
+    // length first rather than calling tre.getCurrentSize() in
+    // case there is no subheader (in which case
+    // subheader.getSubheaderFields() will throw with a
+    // NITRO-specific message).
+    const sys::Uint64_T subheaderLength(subheader.getSubheaderFieldsLength());
+    std::string treTag("");
```

Comment:
Just default construct this and the next string... they'll be empty so no reason to do it explicitly.